test/suites/edgexfoundry/proxy_test.go

```diff
@@ -0,0 +1,99 @@
+package test
+
+import (
+	"edgex-snap-testing/test/utils"
+	"testing"
+
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	snapData        = "/var/snap/edgexfoundry/current/temp"
+	kongAminJwtFile = "/var/snap/edgexfoundry/current/secrets/security-proxy-setup/kong-admin-jwt"
+)
+
+func TestAddProxyUser(t *testing.T) {
+	const (
+		publicKey  = snapData + "/public.pem"
+		privateKey = snapData + "/private.pem"
+	)
+
+	// start clean
+	utils.Exec(t, `sudo rm -rf `+snapData)
+
+	t.Cleanup(func() {
+		utils.Exec(t, `sudo rm -rf `+snapData)
+	})
+
+	// Due to confinement issues when running this test, we write the files to SNAP_DATA
+	utils.Exec(t, `sudo mkdir -p `+snapData)
+
+	// Read the API Gateway token
+	kongAdminJWT, _ := utils.Exec(t, "sudo cat "+kongAminJwtFile)
+	require.NotEmpty(t, kongAdminJWT)
+
+	// Create private and public keys
+	utils.Exec(t, `sudo openssl ecparam -genkey -name prime256v1 -noout -out `+privateKey)
+	utils.Exec(t, `sudo openssl ec -in `+privateKey+` -pubout -out `+publicKey)
+
+	// Use secrets-config to add a user example with id 1000
+	// On success, the above command prints the user id
+	stdout, _ := utils.Exec(t, `
+			sudo edgexfoundry.secrets-config proxy adduser \
+			--token-type jwt --user example --algorithm ES256 \
+			--public_key `+publicKey+` --id 1000 -jwt `+kongAdminJWT)
+	require.Equal(t, "1000\n", stdout)
+
+}
+
+func TestTLSCert(t *testing.T) {
+	// start clean
+	utils.Exec(t, `sudo rm -rf `+snapData)
+
+	t.Cleanup(func() {
+		utils.Exec(t, `sudo rm -rf `+snapData)
+	})
+
+	// Due to confinement issues when running this test, we write the files to SNAP_DATA
+	utils.Exec(t, `sudo mkdir -p `+snapData)
+
+	// Read the API Gateway token
+	kongAminJwt, _ := utils.Exec(t, "sudo cat "+kongAminJwtFile)
+	require.NotEmpty(t, kongAminJwt)
+
+	// Add the certificate, using Kong Admin JWT to authenticate
+	serverKeyFile, serverCertFile := certGenerator()
+	utils.Exec(t, `sudo edgexfoundry.secrets-config proxy tls --incert `+serverCertFile+` --inkey `+serverKeyFile+` --admin_api_jwt `+kongAminJwt)
+
+	code, _ := utils.Exec(t, `curl --show-error --silent --include \
+		--output /dev/null --write-out "%{http_code}" \
+		--cacert `+serverCertFile+` \
+		-X GET 'https://localhost:8443/core-data/api/v2/ping?' \
+		-H "Authorization: Bearer $TOKEN"`)
```

Comment:
```
    log.go:15: [stderr] * TCP_NODELAY set
    log.go:15: [stderr] * Connected to localhost (127.0.0.1) port 8443 (#0)
    log.go:15: [stderr] * ALPN, offering h2
    log.go:15: [stderr] * ALPN, offering http/1.1
    log.go:15: [stderr] * successfully set certificate verify locations:
    log.go:15: [stderr] *   CAfile: /var/snap/edgexfoundry/current/temp/server.crt
    log.go:15: [stderr]   CApath: /etc/ssl/certs
    log.go:15: [stderr] } [5 bytes data]
    log.go:15: [stderr] * TLSv1.3 (OUT), TLS handshake, Client hello (1):
    log.go:15: [stderr] } [512 bytes data]
    log.go:15: [stderr] * TLSv1.3 (IN), TLS handshake, Server hello (2):
    log.go:15: [stderr] { [122 bytes data]
    log.go:15: [stderr] * TLSv1.3 (IN), TLS handshake, Encrypted Extensions (8):
    log.go:15: [stderr] { [19 bytes data]
    log.go:15: [stderr] * TLSv1.3 (IN), TLS handshake, Certificate (11):
    log.go:15: [stderr] { [307 bytes data]
    log.go:15: [stderr] * TLSv1.3 (IN), TLS handshake, CERT verify (15):
    log.go:15: [stderr] { [78 bytes data]
    log.go:15: [stderr] * TLSv1.3 (IN), TLS handshake, Finished (20):
    log.go:15: [stderr] { [52 bytes data]
    log.go:15: [stderr] * TLSv1.3 (OUT), TLS change cipher, Change cipher spec (1):
    log.go:15: [stderr] } [1 bytes data]
    log.go:15: [stderr] * TLSv1.3 (OUT), TLS handshake, Finished (20):
    log.go:15: [stderr] } [52 bytes data]
    log.go:15: [stderr] * SSL connection using TLSv1.3 / TLS_AES_256_GCM_SHA384
    log.go:15: [stderr] * ALPN, server accepted to use h2
    log.go:15: [stderr] * Server certificate:
    log.go:15: [stderr] *  subject: CN=localhost
    log.go:15: [stderr] *  start date: Jun 24 13:39:34 2022 GMT
    log.go:15: [stderr] *  expire date: Mar 20 13:39:34 2025 GMT
    log.go:15: [stderr] *  common name: localhost (matched)
    log.go:15: [stderr] *  issuer: CN=localhost
    log.go:15: [stderr] *  SSL certificate verify ok.
    log.go:15: [stderr] * Using HTTP2, server supports multi-use
    log.go:15: [stderr] * Connection state changed (HTTP/2 confirmed)
    log.go:15: [stderr] * Copying HTTP/2 data in stream buffer to connection buffer after upgrade: len=0
    log.go:15: [stderr] } [5 bytes data]
    log.go:15: [stderr] * Using Stream ID: 1 (easy handle 0x55609ce972f0)
    log.go:15: [stderr] } [5 bytes data]
    log.go:15: [stderr] > GET /core-data/api/v2/ping HTTP/2
    log.go:15: [stderr] > Host: localhost:8443
    log.go:15: [stderr] > user-agent: curl/7.68.0
    log.go:15: [stderr] > accept: */*
    log.go:15: [stderr] > 
    log.go:15: [stderr] { [5 bytes data]
    log.go:15: [stderr] * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
    log.go:15: [stderr] { [265 bytes data]
    log.go:15: [stderr] * TLSv1.3 (IN), TLS handshake, Newsession Ticket (4):
    log.go:15: [stderr] { [265 bytes data]
    log.go:15: [stderr] * old SSL session ID is stale, removing
    log.go:15: [stderr] { [5 bytes data]
    log.go:15: [stderr] * Connection state changed (MAX_CONCURRENT_STREAMS == 128)!
    log.go:15: [stderr] } [5 bytes data]
    log.go:15: [stderr] < HTTP/2 401 
    log.go:15: [stderr] < date: Fri, 24 Jun 2022 13:39:34 GMT
    log.go:15: [stderr] < content-type: application/json; charset=utf-8
    log.go:15: [stderr] < content-length: 26
    log.go:15: [stderr] < x-kong-response-latency: 9
    log.go:15: [stderr] < server: kong/2.6.0
    log.go:15: [stderr] < 
    log.go:15: [stderr] { [26 bytes data]
100    26  100    26    0     0   2166      0 --:--:-- --:--:-- --:--:--  2166
    log.go:15: [stderr] * Connection #0 to host localhost left intact
    log.go:15: [stdout] {"message":"Unauthorized"}
```

Reply:
This looks fine (successful). Note that 401 has nothing to do with TLS. Getting that body means the request was successful. 

401 is because the endpoint is protected and accessing it needs a token, unrelated to this test.

Reply:
Instead of curl and that unreadable output, you can try go's http.Get.

Reply:
Thanks for the suggestion. I have tried sending an HTTP request with a certificate in golang:
```
	caCert, err := ioutil.ReadFile(caCertFile)
	if err != nil {
		t.Log(err)
	}
	caCertPool := x509.NewCertPool()
	caCertPool.AppendCertsFromPEM(caCert)

	client := &http.Client{
		Transport: &http.Transport{
			TLSClientConfig: &tls.Config{
				RootCAs: caCertPool,
			},
		},
	}

	resp, err := client.Get("https://localhost:8443/core-data/api/v2/ping")
	if err != nil {
		t.Log(err)
	}
	t.Log(resp.StatusCode)
	require.Equal(t, 200, resp.StatusCode)
```
error:
```
Get "https://localhost:8443/core-data/api/v2/ping": x509: certificate relies on legacy Common Name field, use SANs instead
```

Reply:
> This looks fine (successful). Note that 401 has nothing to do with TLS. Getting that body `{"message":"Unauthorized"}` means the request was successful.
> 
> 401 is because the endpoint is protected and accessing it needs a token, unrelated to this test.

What is the exit code when using curl?

Reply:
It is 200 now after using correct token

Reply:
I didn't know you can use the admin token to call an API. But as I said that isn't relevant to this test. Status 401 was the correct and expected response. My question was about the exit code of curl. If exit code is 0 and status is 401, the test should pass.